src/Dapr.Client/DaprClientGrpc.cs

```diff
@@ -294,10 +355,100 @@ public override async ValueTask<IReadOnlyList<BulkStateItem>> GetBulkStateAsync(
+
+                var response = await grpcCall.ResponseAsync;
+                var responseHeaders = await grpcCall.ResponseHeadersAsync;
+                var trailers = grpcCall.GetTrailers();
+                var grpcStatus = grpcCall.GetStatus();
+
+                var headers = grpcCall.ResponseHeadersAsync.Result.ToDictionary(kv => kv.Key, kv => kv.ValueBytes);
+
+                    invokeResponse.Body = response.Data.Value.IsEmpty ? default : TypeConverters.FromAny<TResponse>(response.Data, this.jsonSerializerOptions);
+                    invokeResponse.Headers = headers;
+                    invokeResponse.Trailers = grpcCall.GetTrailers().ToDictionary(kv => kv.Key, kv => kv.ValueBytes);
+
+                if (IsResponseFromHttpCallee(headers))
+                {
+                    invokeResponse.HttpStatusCode = (HttpStatusCode)System.Enum.Parse(typeof(HttpStatusCode), Encoding.UTF8.GetString(headers[daprHTTPStatusHeader], 0, headers[daprHTTPStatusHeader].Length));
+                    invokeResponse.ContentType = Constants.ContentTypeApplicationJson;
+                }
+                else
+                {
+                    // Response is grpc
+                    invokeResponse.GrpcStatusInfo = new GRPCStatusInfo(grpcStatus.StatusCode, grpcStatus.Detail);
+                    invokeResponse.ContentType = Constants.ContentTypeApplicationGrpc;
+                }
+            }
+            catch (RpcException ex)
+            {
+                throw ProcessRpcException(ex, invokeResponse);
+            }
+
+            return invokeResponse;
+        }
+
+        public static ServiceInvocationException<TRequest, TResponse> ProcessRpcException<TRequest, TResponse>(RpcException ex, ServiceInvocationResponse<TRequest, TResponse> invokeResponse)
+        {
+            var entry = ex.Trailers.Get(grpcStatusDetails);
+            if (entry != null)
+            {
+                var status = Google.Rpc.Status.Parser.ParseFrom(entry.ValueBytes);
+                foreach (var detail in status.Details)
+                {
+                    if (Google.Protobuf.WellKnownTypes.Any.GetTypeName(detail.TypeUrl) == grpcErrorInfoDetail)
+                    {
+                        var rpcError = detail.Unpack<Google.Rpc.ErrorInfo>();
+                        var grpcStatusCode = (StatusCode) System.Enum.ToObject(typeof(StatusCode), status.Code);
+                        invokeResponse.GrpcStatusInfo = new GRPCStatusInfo(grpcStatusCode, status.Message, Convert.ToInt32(rpcError.Metadata[daprErrorInfoHTTPCodeMetadata]), rpcError.Metadata[daprErrorInfoHTTPErrorMetadata]);
+                    }
+                }
+            }
+
+            return new ServiceInvocationException<TRequest, TResponse>($"Exception while invoking {invokeResponse.Request.MethodName} on appId:{invokeResponse.Request.AppId}", ex, invokeResponse);
```

Comment:
Let's think about this more. I'm not sure we want to make the exception a generic type. 

Writing error handling that's general purpose doesn't really work when the exception has generics in it, because you'd have to predict all of the replacements of `TRequest, TResponse`. 

Generic exception are exceedingly rare for this reason.

Reply:
Valid point but we cannot include the Response in the exception then.